Better dbt packages version logging to aid in upgrading outdated packages

sungchun12 opened the pull request:
resolves #3560 #3763

<!---
  Include the number of the issue addressed by this PR above if applicable.
  PRs for code changes without an associated issue *will not be merged*.
  See CONTRIBUTING.md for more information.

  Example:
    resolves #1234
-->


### Description

<!--- Describe the Pull Request here -->
- For publically available dbt packages through the [package registry](https://hub.getdbt.com/), provide clear dbt package versions being installed vs. the latest version where applicable.
- Guides the developer to update older dbt packages
```yaml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.7.0
  - package: dbt-labs/codegen
    version: 0.4.0
  - package: calogica/dbt_expectations
    version: 0.4.0
```
```bash
# display which specific packages are available for upgrade
# display everything that is up to date
$ dbt deps
Running with dbt=0.21.0-b1
Installing dbt-labs/dbt_utils@0.7.0
  Installed from version 0.7.0
  Latest hub registry version 0.7.1
Installing dbt-labs/codegen@0.4.0
  Installed from version 0.4.0
  Up to date!
Installing calogica/dbt_expectations@0.4.0
  Installed from version 0.4.0
  Latest hub registry version 0.4.1
Installing calogica/dbt_date@0.4.0
  Installed from version 0.4.0
  Up to date!

Upgrades available for: ['dbt-labs/dbt_utils', 'calogica/dbt_expectations']
```
![image](https://user-images.githubusercontent.com/19176976/130137105-1070c2de-89b7-4c37-9267-9c2fc53aa22d.png)


```yaml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.7.1
  - package: dbt-labs/codegen
    version: 0.4.0
  - package: calogica/dbt_expectations
    version: 0.4.1
```
```bash
$ dbt deps
Running with dbt=0.21.0-b1
Installing dbt-labs/dbt_utils@0.7.1
  Installed from version 0.7.1
  Up to date!
Installing dbt-labs/codegen@0.4.0
  Installed from version 0.4.0
  Up to date!
Installing calogica/dbt_expectations@0.4.1
  Installed from version 0.4.1
  Up to date!
Installing calogica/dbt_date@0.4.0
  Installed from version 0.4.0
  Up to date!
```
![image](https://user-images.githubusercontent.com/19176976/130141160-2bd7dc6a-c5ad-4c7b-a8a1-7f3661d7793f.png)

**Out of scope**: git-installed and local dbt packages

**Notes**:
- Tested `get_version_latest()` method call in` test/unit/test_deps.py`. This is only relevant for hub registry packages

**Open Questions**
- Test logging outputs in `test/rpc/test_deps.py`? I looked across several test files relevant to `deps` and didn't see any related to validating logging outputs. Can you provide guidance on where to best insert logging test cases/is it needed here given we aren't testing that today?

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

sungchun12 commented:
**Notes**:
- Test method calls in test/unit/test_deps.py
- Test logging outputs in test/rpc/test_deps.py
- Create separate method in class instead of bloating if/elif conditions?

joellabes commented:
Love it! Thanks @sungchun12 😍 

My only nitpick is around wording where updates are available. What if it was instead 

```
# display which specific packages are available for upgrade
# display everything that is up to date
$ dbt deps
Running with dbt=0.21.0-b1
Installing dbt-labs/dbt_utils@0.7.0
  Installed from version 0.7.0
  Update available: version 0.7.1
Installing dbt-labs/codegen@0.4.0
  Installed from version 0.4.0
  Up to date!
Installing calogica/dbt_expectations@0.4.0
  Installed from version 0.4.0
  Update available: version 0.4.1
Installing calogica/dbt_date@0.4.0
  Installed from version 0.4.0
  Up to date!

Updates available for: ['dbt-labs/dbt_utils', 'calogica/dbt_expectations']
```

I.e. 
- Replace `Latest hub registry` with `Update available:` to make it a) more glanceable and b) more actionable, 
- Replace `Upgrades available` with `Updates available`. I feel less strongly about this one, but it's more in line with the `Up to date!` messaging.

The only other thing I'd consider is making the "OK, and what should I do about it?" more clear, which could be as simple as appending this to the bottom of the log when updates are available: `Update your versions in packages.yml, then run dbt deps`. If you installed a package then you probably know how to update it, but I'm thinking of users who had the package installed for them and are coming in fresh.

sungchun12 commented:
@joellabes Thanks for the clear feedback! I'll make changes to the logging output!

Is there a reason you want to remove `hub` language from the terminal output? I recommend we keep it, so it's clear that the updates are specific to hub registry packages. I want to avoid users getting confused, especially if it's their first time with `dbt deps`, that this logging applies to git and local packages at a glance.

joellabes commented:
Fair point! I primarily wanted a clear action item `Update available` and thought I should try to trim down the total word count for the message. 

How about something like `Update available from hub: version 0.7.1`? My only thing then is I wonder whether people know that the package hub is called a hub - I don't think it really matters if they don't though, you're right that clarifying which packages are eligible for this check is very relevant. 

Out of curiosity, is anything shown for non-hub packages? E.g. if you are doing a direct git installation does it say something like `Update information unavailable for git-based installations`?

sungchun12 commented:
@joellabes I updated the logging outputs based on your feedback. I also updated the `Updates available` section to be specific to `hub` packages. No `up to date` language will appear respective to git and local packages. The logic implemented specifically checks if the package is from hub before doing all these version checks. 

I recommend against outputting extraneous logging related to git and local packages as that can be deduced from the lack of version language in the logs. If you want, we can add this disclaimer specifically to the dbt docs repo. @jtcohen6 Do you have a recommendation either way?

Updated examples below.

```yaml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.7.0
  - package: dbt-labs/codegen
    version: 0.4.0
  - package: calogica/dbt_expectations
    version: 0.4.0
  - git: https://github.com/dbt-labs/dbt-audit-helper.git
    revision: 0.4.0
  - git: "https://github.com/dbt-labs/dbt-labs-experimental-features" # git URL
    subdirectory: "materialized-views" # name of subdirectory containing `dbt_project.yml`
    revision: 0.0.1
```

```txt
Running with dbt=0.21.0-b1
Installing dbt-labs/dbt_utils@0.7.0
  Installed from version 0.7.0
  Update available from hub: version 0.7.1
Installing dbt-labs/codegen@0.4.0
  Installed from version 0.4.0
  Up to date!
Installing calogica/dbt_expectations@0.4.0
  Installed from version 0.4.0
  Update available from hub: version 0.4.1
Installing https://github.com/dbt-labs/dbt-audit-helper.git@0.4.0
  Installed from revision 0.4.0
Installing https://github.com/dbt-labs/dbt-labs-experimental-features@0.0.1
  Installed from revision 0.0.1
   and subdirectory materialized-views
Installing calogica/dbt_date@0.4.0
  Installed from version 0.4.0
  Up to date!

Updates available from hub: ['dbt-labs/dbt_utils', 'calogica/dbt_expectations']                 
Update your versions in packages.yml, then run dbt deps
```

![image](https://user-images.githubusercontent.com/19176976/130456650-0ff39ccc-7f78-4835-bd56-586757236723.png)

```yaml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.7.1
  - package: dbt-labs/codegen
    version: 0.4.0
  - package: calogica/dbt_expectations
    version: 0.4.1
  - git: https://github.com/dbt-labs/dbt-audit-helper.git
    revision: 0.4.0
  - git: "https://github.com/dbt-labs/dbt-labs-experimental-features" # git URL
    subdirectory: "materialized-views" # name of subdirectory containing `dbt_project.yml`
    revision: 0.0.1
```

```txt
Running with dbt=0.21.0-b1
Installing dbt-labs/dbt_utils@0.7.1
  Installed from version 0.7.1
  Up to date!
Installing dbt-labs/codegen@0.4.0
  Installed from version 0.4.0
  Up to date!
Installing calogica/dbt_expectations@0.4.1
  Installed from version 0.4.1
  Up to date!
Installing https://github.com/dbt-labs/dbt-audit-helper.git@0.4.0
  Installed from revision 0.4.0
Installing https://github.com/dbt-labs/dbt-labs-experimental-features@0.0.1
  Installed from revision 0.0.1
   and subdirectory materialized-views
Installing calogica/dbt_date@0.4.0
  Installed from version 0.4.0
  Up to date!
```

![image](https://user-images.githubusercontent.com/19176976/130456811-daec8f66-ef43-477d-8e13-f4c969198c44.png)

joellabes commented:
Yep this all looks great 😍 happy for the git etc ones to have no logging, but would like that to be called out as a one-liner in the docs

sungchun12 commented:
Docs draft pull request [#785](https://github.com/dbt-labs/docs.getdbt.com/pull/785)

sungchun12 commented on core/dbt/deps/registry.py:
```diff
@@ -120,6 +127,8 @@ def resolved(self) -> RegistryPinnedPackage:
             raise DependencyException(new_msg) from e
 
         available = registry.get_available_versions(self.package)
+        available_latest = max(available)
```
I changed things up by using the built-in `StrictVersion` package. It looks like we enforce python version semantics, so this should work. If we expect to have looser semantics, we can use `LooseVersion `

```python
def filter_installable(
        versions: List[str],
        install_prerelease: bool
) -> List[str]:
    installable = []
    for version_string in versions:
        version = VersionSpecifier.from_version_string(version_string)
        if install_prerelease or not version.prerelease:
            installable.append(version_string)
    installable.sort(key=StrictVersion)
    return installable
```

![image](https://user-images.githubusercontent.com/19176976/131365645-fe7dc9ea-fb33-43c0-9850-750079ce5556.png)

sungchun12 commented:
**Ready for Another Review!**
- Replied inline to above comments
- Updated `test__filter_installable` with a raw, unsorted list to verify filter_installable can properly sort versions with or without prereleases
- Updated `get_version_latest` unit tests with: `test_get_version_latest_prelease_true` and `test_get_version_latest_prelease_false`

Examples below.
![image](https://user-images.githubusercontent.com/19176976/131369224-fc70d74c-208c-430c-bf39-9e0e0dab6e10.png)
![image](https://user-images.githubusercontent.com/19176976/131369268-6d408695-a30f-46f1-81af-fe31f863985a.png)

![image](https://user-images.githubusercontent.com/19176976/131369458-99318819-1f6e-4460-9e82-057f477badb8.png)
![image](https://user-images.githubusercontent.com/19176976/131369472-b9e5b956-d58a-4857-9707-aee2ff1730a6.png)

![image](https://user-images.githubusercontent.com/19176976/131369698-519b993d-be70-4d69-a7d1-e2f791359632.png)
![image](https://user-images.githubusercontent.com/19176976/131369725-b7cc4de1-9c44-4f98-984b-0b673e8f1500.png)
![image](https://user-images.githubusercontent.com/19176976/131369827-43e8e7f8-ddc2-4326-a5e0-d26b3cf3c9a8.png)
![image](https://user-images.githubusercontent.com/19176976/131369859-5ffa40b3-5cd6-4127-9abb-7eb8be078898.png)

leahwicz commented on core/dbt/task/deps.py:
```diff
@@ -56,19 +56,36 @@ def run(self):
+                        packages_to_upgrade.append(package_name)
+                        logger.info('  Updated version available: {}',
+                                    version_latest)
+                    elif version_latest == version:
```
Nit: this can be an `else` since there is no other possible state (it can't be not equal to `version` and equal to `version`

sungchun12 replied:
Great call out! Made the change :)

jtcohen6 reviewed (approved):
@sungchun12 Amazing work!! Thanks for the contribution, and for seeing it through. It's a pleasure to merge this in.

sungchun12 commented:
@jtcohen6 I was doing some dbt development and noticed that `StrictVersion` does not allow package versions like `0.2.1-pre`. I'm going to open another pull request and reference this one for an audit trail. Let me know if you want me to document an issue as well!
![image](https://user-images.githubusercontent.com/19176976/131867998-c896c11d-0610-4b42-95a5-25e3498c09ab.png)

It should require a simple fix with `LooseVersion`. 
![image](https://user-images.githubusercontent.com/19176976/131868291-80ef47b3-5afc-477a-b086-9e153e1d26b6.png)
